### ISSUE TITLE
Support Python 12

### ISSUE THREAD

**Issue by markstur** [instructlab/instructlab#1463]
**Is your feature request related to a problem? Please describe.**

Support for Python 12 is being withdrawn in https://github.com/instructlab/instructlab/pull/1461 for reasons.

**Describe the solution you'd like**

This will need to be resolved so py12 support can be brought back.  It's the only real active Python version right now and will be popular for the next few years.

**Additional context**

**Comment by tiran** [comment]
@markstur I have fixed the subject and text of this ticket. It's "Python 3.12", not "Python 12".

The lack of 3.12 support is outside our control. Python 3.12 is currently not supported, because several dependencies and features are not compatible with 3.12. This includes Torch compile / Triton, vLLM, Ray, and maybe more. I expect that PyTorch and the rest of the stack is going to support 3.12 by the end of the year.

There is nothing wrong with 3.11. Fedora, CentOS, and RHEL come with Python 3.11. CPython upstream supports 3.11 until October 2027. We are good for a couple of years.

**Comment by booxter** [comment]
There is nothing wrong about not supporting python3.12 for now. But there's one issue: `tox` doesn't know that python3.12 is not supported and - if installed - will try to use this version to initialize an environment - and will fail.

I don't think `tox` supports capping python version. We could add `basepython` to main environment pointing to `python3.11`, but that will enforce this version for all environments. Not sure we'd want that.

**Comment by tiran** [comment]
Yes, our current `tox.ini` does not play well with the new requirements. `[testenv] basepython = python3.11` does not work, either. It breaks `[testenv:py310-*]` environments. PR #1526 fixes the problem and several other smaller issues.

**Comment by nickschuetz** [comment]
> There is nothing wrong about not supporting python3.12 for now. But there's one issue: `tox` doesn't know that python3.12 is not supported and - if installed - will try to use this version to initialize an environment - and will fail.
> 
> I don't think `tox` supports capping python version. We could add `basepython` to main environment pointing to `python3.11`, but that will enforce this version for all environments. Not sure we'd want that.
> 
> This can be achieved locally with: `tox -x testenv.basepython=python3.11 -e mypy`

Note that the Mac developer experience will be tainted in that 3.12 is the default python brew install. Not a show stopper obviously but a kink in the road for anyone who is looking to hit the ground running on that platform.

The same goes for Fedora 40.

**Comment by nickschuetz** [comment]
> For the record, Fedora supports a dozen Python versions and implementation. Python 3.11 is one of them.

However, the default `python` install is 3.12. That's the point I'm trying to make.

**Comment by fabiendupont** [comment]
PyTorch 2.4.0 supports Python 3.12, as well as vLLM v0.5.4+ and main branch of ODH fork of vLLM. It would be great to test instructlab with 3.12 to see if there's any major issue with the dependencies.

**Comment by nathan-weinberg** [comment]
_dusts off_

**Comment by nathan-weinberg** [comment]
Open Questions:
- Do we drop Python 3.10 support once we start supporting 3.12?
- What is our general Python support policy? Historically we supported Python 3.9 through 3.12 before having to restrict to Python 3.10 and 3.11. Do we want to continue only supporting two Python versions or go to supporting three? Maybe more?

**Comment by tiran** [comment]
@ktam3 Could you please make this a requirement for 0.26 release? One of our AI accelerator stack is dropping support for Python 3.11 in the next couple of weeks.

@nathan-weinberg The decision is entirely up to you. How much time, energy and CI resources does the InstructLab team have to verify the code on multiple Python versions? How many users are installing InstructLab on Python 3.10? https://pypistats.org/ and https://distrowatch.com/ are useful to figure out package usage and default Python version of distros.

**Comment by RobotSail** [comment]
@tiran I'm definitely not using Python 3.10. And now that we have Liger kernels added in the upstream, we should aim to deprecate the use of GPTDolomite. The only thing is that there are some setups which rely on GPTDolomite in order to train longer sequences on A100 & H100 GPUs. So some work will need to be done to fully realize the memory gains that Liger offers. 

Once we are able to train the same length or longer sequences, we should be all set for this.

**Comment by ktam3** [comment]
looping in @courtneypacheco and @ktdreyer here

**Comment by ktam3** [comment]
GPTDolomite will need to be deprecated first before we bump to 3.12

Per oleg: 

> Last night I was training with deepspeed and was able to get around the same memory usage with Liger as with GPTDolomite, so I think it will be okay to drop GPTDolomite in this release. If we find that we still need GPTDolomite then we can revisit that. But yes let's definitely move off of these old versions

I'm not sure if there are other items blocking this bump, but if not, we can plan to do this for the next release cycle

**Comment by booxter** [comment]
Before there can be progress switching jobs to 3.12+, some consolidated decision should be made about how we support multiple versions. I posted this document: https://github.com/instructlab/dev-docs/pull/200 but it doesn't see much traction. In my mind, it should have been merged; then we would propose a separate update to it bumping the version for the preferred Python from 3.11 to 3.12, which would then be a signal to go across all repos and change defaults to 3.12.

**Comment by booxter** [comment]
dolomite removal: https://github.com/instructlab/instructlab/pull/3491
